onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -77,84 +80,57 @@ bool ReshapeFusion::Fuse_Subgraph1(Node& reshape, Graph& graph, const logging::L
-    return false;
-  }
-
-  const Node& unsqueeze_2 = edges[0]->GetNode();
-  const Node& gather_2 = edges[1]->GetNode();
-  const Node& shape_2 = edges[2]->GetNode();
-  if (unsqueeze_2.GetOutputEdgesCount() != 1 || gather_2.GetOutputEdgesCount() != 1 || shape_2.GetOutputEdgesCount() != 1) {
-    return false;
-  }
+  std::vector<int64_t> shape_value;
+  shape_value.reserve(concat_input_count);
+  // Used to keep the following nodes in the order of their potential removal.
+  enum class NodeType { Unsqueeze, Gather, Shape };
+  std::set<std::pair<NodeType, NodeIndex>> candidates_for_removal;
+  for (int i = 0; i < concat_input_count; ++i) {
+    // First check if the i-th argument is an initializer.
+    // We do not check whether the initializer is constant.
+    // Some model uses constant initializer and some does not.
+    // Here we assume that no one will override the initializer using graph input.
+    if (optimizer_utils::AppendTensorFromInitializer(graph, *(concat.InputDefs()[i]), shape_value)) {
+      continue;
+    }
 
-  if (graph_utils::GetInputNode(shape_2, 0) != p_root) {
-    return false;
-  }
+    // Try to find path [Root] --> Shape --> Gather(indices=i) --> Unsqueeze (axes=0) --> Concat [input i]
+    std::vector<graph_utils::EdgeEndToMatch> parent_path{
+        {0, i, "Unsqueeze", {1, 11}, kOnnxDomain},
+        {0, 0, "Gather", {1, 11}, kOnnxDomain},
+        {0, 0, "Shape", {1}, kOnnxDomain}};
 
-  if (!(graph_utils::GetRepeatedNodeAttributeValues(unsqueeze_2, "axes", axes) && axes.size() == 1 && axes[0] == 0)) {
-    return false;
-  }
+    std::vector<const Node::EdgeEnd*> edges;
+    if (!graph_utils::FindPath(concat, true, parent_path, edges, logger)) {
+      return false;
+    }
 
-  if (!optimizer_utils::IsInitializerWithExpectedValue(graph, *(gather_2.InputDefs()[1]), int64_t(1), false)) {
-    return false;
-  }
+    const Node& unsqueeze = edges[0]->GetNode();
+    const Node& gather = edges[1]->GetNode();
+    const Node& shape = edges[2]->GetNode();
 
-  // Compose the shape value input for reshape op.
-  std::vector<int64_t> shape_value = {0, 0};
+    if (graph_utils::GetInputNode(shape, 0) != p_root) {
+      return false;
+    }
 
-  // We do not check whether the initializer is constant.
-  // Some model uses constant initializer and some does not.
-  // Here we assume that no one will override the initializer using graph input.
-  if (!optimizer_utils::AppendTensorFromInitializer(graph, *(concat.InputDefs()[2]), shape_value)) {
-    return false;
-  }
+    std::vector<int64_t> axes;
+    if (!(graph_utils::GetRepeatedNodeAttributeValues(unsqueeze, "axes", axes) && axes.size() == 1 && axes[0] == 0)) {
+      return false;
+    }
 
-  if (concat_input_count > 3) {
-    if (!optimizer_utils::AppendTensorFromInitializer(graph, *(concat.InputDefs()[3]), shape_value)) {
+    if (!optimizer_utils::IsInitializerWithExpectedValue(graph, *(gather.InputDefs()[1]), int64_t(i), false)) {
```

Comment:
The expected value should be shape_value.size() instead of i because the tensor appended in line 97 could have multiple elements. @MaximKalininMS, could you provide a fix?

Reply:
Thanks for catching it! Sent https://github.com/microsoft/onnxruntime/pull/3721